accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,671 @@
++       public static Type MakeFunctionPointerType(Type[] functionPointerArguments);
+    }
+}
+```
+### Examples
+```cs
+// Simple managed function pointer
+Type type = typeof(delegate*<int, bool>);
+Debug.Assert(type.IsFunctionPointer == true);
+Debug.Assert(type.IsUnmanagedFunctionPointer == false);
+Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));
+Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
+Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));
+```
+
+```cs
+// Simple unmanaged function pointer
+Type type = typeof(delegate* unmanaged[Cdecl, MemberFunction]<int, bool>);
+Debug.Assert(type.IsFunctionPointer == true);
+Debug.Assert(type.IsUnmanagedFunctionPointer == true);
+Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));
+
+Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
+Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));
+```
+
+## GetModified*Type() methods for `FieldInfo`, `PropertyInfo` and `ParameterInfo`
+If the type returned from the corresponding `FieldInfo.FieldType`, `PropertyInfo.PropertyType` or `ParameterInfo.ParameterType` is not a function pointer, then the return type is `null`.
+
+**discussion: we could also return the underlying, unmodified function pointer type. This requires the caller to always check `IsModifiedType` before using the members like `GetFunctionPointerCallingConventions()`.**
+
+However, if the return type is a function pointer, then a "modified type" is returned which is a wrapper over the type returned from `FieldType \ PropertyType \ Parametertype`. 
+
+Each type returned from `GetFunctionPointerParameterTypes()`, whether a primitive such as `String` or a user-defined class, is considered a modified type that contains any custom modifiers present on the parameter declaration, such as `ref\out\in`. If it contains no modifiers, it is essentially an empty type with just a reference to its underlying type, such as `String`.
+
+From a modified parameter type, the custom modifiers can be accessed via `GetOptionalCustomModifiers()` and `GetRequiredCustomModifiers()`.
+
+```diff
+namespace System.Reflection
+{
+    public abstract class FieldInfo
+    {
++       public virtual Type? GetModifiedFieldType() => null;
+        // or "GetFieldModifiedType"?
+    }
+
+    public abstract class PropertyInfo
+    {
++       public virtual Type? GetModifiedPropertyType() => null;
+    }
+
+    public abstract class ParameterInfo
+    {
++       public virtual Type? GetModifiedParameterType() => null;
+    }
+}
+```
+
+### Examples
+```c#
+public class Holder
+{
+    public unsafe delegate*<in int, out int, void> _field;
+}
+
+FieldInfo fieldInfo = typeof(Holder).GetField("_field");
+
+// Get the modified type from the field; this Type is like fieldInfo.FieldType but it includes custom mods
+Type fnType = fieldInfo.GetModifiedFieldType();
```

Comment:
```suggestion
Type fnType = fieldInfo.GetModifiedFieldType()!;
```
What is the right way to deal with `GetModifiedFieldType` possibly returning null? Add `!` everywhere?

Reply:
I updated the design to return non-null. By doing this, we may return an unmodified type. The design was updated to state that it really shouldn't matter to the caller since an unmodified and modified type have the same semantics for things like GetRequiredCustomModifiers() where we just return an empty array (and don't throw for example).

Reply:
To keep consistent semantics, `GetModified*Type()` will always return a "modified type" even when it could return a "runtime type". This is because a modified type (for the runtime anyway) will not derive from the internal `RuntimeType` and instead derive from `TypeDelegator` which does not derive from `RuntimeType` and we want to ensure the same semantics (throw or allow) for the methods that only take `RuntimeType` (performed when validating parameters).

For example, if we don't "need to" create a modified type in `GetModified*Type()` (because there are no modifiers in that type or any nested types), and instead create a runtime type (which I'm proposing we DON'T want to do), the user may start passing those to methods that only take runtime types which will work fine until  `GetModified*Type()` needs to return a modified type which fails the "pit of success" test.